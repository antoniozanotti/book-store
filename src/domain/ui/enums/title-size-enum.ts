export enum TitleSizeEnum {
  "small" = "small",
  "medium" = "medium",
  "large" = "large",
  "x_large" = "x_large",
}
