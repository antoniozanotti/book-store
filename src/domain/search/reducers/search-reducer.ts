import { SearchType } from "../types/search-type";
import { SearchActionType } from "../types/search-action-type";
import { SearchActionEnum } from "../enums/search-action-enum";
import { DEFAULT_ORDER_BY } from "../consts";
import { MAX_PRICE, MAX_YEAR, MIN_PRICE, MIN_YEAR } from "@/const/global";

export function searchReducer(
  search: SearchType,
  action: SearchActionType
): SearchType {
  switch (action.type) {
    case SearchActionEnum.UPDATED_KEYWORD:
      return handleUpdatedKeyword(search, action);
    case SearchActionEnum.UPDATED_ORDERBY:
      return handleUpdatedOrderBy(search, action);
    case SearchActionEnum.UPDATED_CATEGORIES:
      return handleUpdatedCategories(search, action);
    case SearchActionEnum.UPDATED_PRICE:
      return handleUpdatedPrice(search, action);
    case SearchActionEnum.UPDATED_YEAR:
      return handleUpdateYear(search, action);
    default:
      throw Error("Unknown action: " + action.type);
  }
}

function handleUpdatedKeyword(search: SearchType, action: SearchActionType) {
  return {
    ...search,
    keyword: action.payload.keyword,
    categories: [""],
    minprice: MIN_PRICE,
    maxprice: MAX_PRICE,
    minyear: MIN_YEAR,
    maxyear: MAX_YEAR,
    orderby: DEFAULT_ORDER_BY,
  };
}

function handleUpdatedOrderBy(search: SearchType, action: SearchActionType) {
  return { ...search, orderby: action.payload.orderby };
}

function handleUpdatedCategories(search: SearchType, action: SearchActionType) {
  return { ...search, categories: action.payload.categories };
}

function handleUpdatedPrice(search: SearchType, action: SearchActionType) {
  return {
    ...search,
    minprice: action.payload.minprice,
    maxprice: action.payload.maxprice,
  };
}

function handleUpdateYear(search: SearchType, action: SearchActionType) {
  return {
    ...search,
    minyear: action.payload.minyear,
    maxyear: action.payload.maxyear,
  };
}