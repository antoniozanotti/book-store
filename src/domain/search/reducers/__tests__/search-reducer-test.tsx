import { searchMock } from "../../__tests__/search-mock";
import { OrderByEnum } from "../../enums/order-by-enum";
import { SearchActionEnum } from "../../enums/search-action-enum";
import { SearchType } from "../../types/search-type";
import { searchReducer } from "../search-reducer";

describe("About", () => {
  it("should update keyword", async () => {
    let search: SearchType = searchMock;

    search = searchReducer(search, {
      type: SearchActionEnum.UPDATED_KEYWORD,
      payload: { keyword: "Test" },
    });

    expect(search).toMatchSnapshot();
  });

  it("should update orderby", async () => {
    let search: SearchType = searchMock;

    search = searchReducer(search, {
      type: SearchActionEnum.UPDATED_ORDERBY,
      payload: { orderby: OrderByEnum.PRICE },
    });

    expect(search).toMatchSnapshot();
  });

  it("should update categories", async () => {
    let search: SearchType = searchMock;

    search = searchReducer(search, {
      type: SearchActionEnum.UPDATED_CATEGORIES,
      payload: { categories: ["a", "b"] },
    });

    expect(search).toMatchSnapshot();
  });

  it("should update price", async () => {
    let search: SearchType = searchMock;

    search = searchReducer(search, {
      type: SearchActionEnum.UPDATED_PRICE,
      payload: { minprice: 10, maxprice: 20 },
    });

    expect(search).toMatchSnapshot();
  });

  it("should update year", async () => {
    let search: SearchType = searchMock;

    search = searchReducer(search, {
      type: SearchActionEnum.UPDATED_YEAR,
      payload: { minyear: 2005, maxyear: 2010 },
    });

    expect(search).toMatchSnapshot();
  });

  it("should throw error", async () => {
    let search: SearchType = searchMock;

    const error = () => {
      searchReducer(search, {
        type: "error",
      });
    };

    expect(error).toThrow(Error);
  });
});
