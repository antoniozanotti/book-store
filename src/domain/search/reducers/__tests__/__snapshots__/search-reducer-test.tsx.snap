// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`About should update categories 1`] = `
{
  "categories": [
    "a",
    "b",
  ],
  "keyword": "",
  "maxprice": 499,
  "maxyear": 2024,
  "minprice": 20,
  "minyear": 1990,
  "orderby": "name",
}
`;

exports[`About should update keyword 1`] = `
{
  "categories": [
    "",
  ],
  "keyword": "Test",
  "maxprice": 499,
  "maxyear": 2024,
  "minprice": 20,
  "minyear": 1990,
  "orderby": "name",
}
`;

exports[`About should update orderby 1`] = `
{
  "categories": [],
  "keyword": "",
  "maxprice": 499,
  "maxyear": 2024,
  "minprice": 20,
  "minyear": 1990,
  "orderby": "price",
}
`;

exports[`About should update price 1`] = `
{
  "categories": [],
  "keyword": "",
  "maxprice": 20,
  "maxyear": 2024,
  "minprice": 10,
  "minyear": 1990,
  "orderby": "name",
}
`;

exports[`About should update year 1`] = `
{
  "categories": [],
  "keyword": "",
  "maxprice": 499,
  "maxyear": 2010,
  "minprice": 20,
  "minyear": 2005,
  "orderby": "name",
}
`;
